File: app/lib/utils/hard_secret_detector.dart
class HardSecretDetection {
  final String category;
  final int start;
  final int end;

  const HardSecretDetection({required this.category, required this.start, required this.end});
}

class HardSecretDetector {
  static final List<({String category, RegExp pattern})> _patterns = [
    (
      category: 'private_key',
      pattern: RegExp(r'-----BEGIN\s+(?:RSA\s+|EC\s+|OPENSSH\s+|DSA\s+)?PRIVATE\s+KEY-----', caseSensitive: false),
    ),
    (
      category: 'database_url',
      pattern: RegExp(
        r"""\b(?:postgres|postgresql|mysql|mongodb(?:\+srv)?|redis)://[^\s<>"']{12,}""",
        caseSensitive: false,
      ),
    ),
    (
      category: 'api_key',
      pattern: RegExp(r'\b(?:sk[-_][A-Za-z0-9_-]{16,}|ghp_[A-Za-z0-9_]{20,}|AKIA[0-9A-Z]{16})\b', caseSensitive: true),
    ),
    (
      category: 'token',
      pattern: RegExp(
        r"""\b(?:(?:access[_-]?|auth[_-]?|refresh[_-]?)?token\b\s*[:=]\s*["']?|authorization\b\s*:\s*bearer\s+|bearer\b\s+)[A-Za-z0-9._~+/=-]{16,}""",
        caseSensitive: false,
      ),
    ),
    (
      category: 'password',
      pattern: RegExp(r"""\b(?:password|passwd|pwd)\b\s*[:=]\s*["']?[^"'\s]{8,}""", caseSensitive: false),
    ),
    (
      category: 'cookie',
      pattern: RegExp(
        r"""\b(?:cookie|session(?:id)?|sid)\b\s*[:=]\s*["']?[A-Za-z0-9._~+/=-]{16,}""",
        caseSensitive: false,
      ),
    ),
  ];

  static List<HardSecretDetection> detections(String text) {
    final hits = <HardSecretDetection>[];
    for (final candidate in _patterns) {
      for (final match in candidate.pattern.allMatches(text)) {
        hits.add(HardSecretDetection(category: candidate.category, start: match.start, end: match.end));
      }
    }
    hits.sort((a, b) {
      final byStart = a.start.compareTo(b.start);
      if (byStart != 0) return byStart;
      return a.category.compareTo(b.category);
    });
    return hits;
  }

  static bool contains(String text) => detections(text).isNotEmpty;

  static List<String> categories(String text) {
    final values = detections(text).map((detection) => detection.category).toSet().toList();
    values.sort();
    return values;
  }
}

File: desktop/macos/Desktop/Sources/HardSecretDetector.swift
import Foundation

struct HardSecretDetection: Equatable {
  let category: String
  let range: Range<String.Index>
}

enum HardSecretDetector {
  private struct Pattern {
    let category: String
    let regex: NSRegularExpression
  }

  private static let patterns: [Pattern] = [
    Pattern(
      category: "private_key",
      regex: try! NSRegularExpression(
        pattern: #"-----BEGIN [A-Z ]*PRIVATE KEY-----[\s\S]*?-----END [A-Z ]*PRIVATE KEY-----"#,
        options: []
      )
    ),
    Pattern(
      category: "private_key",
      regex: try! NSRegularExpression(
        pattern: #"-----BEGIN [A-Z ]*PRIVATE KEY-----"#,
        options: []
      )
    ),
    Pattern(
      category: "database_url",
      regex: try! NSRegularExpression(
        pattern: #"\b[a-z][a-z0-9+.-]*://[^/\s:@]+:[^@\s]+@[^/\s]+[^\s]*"#,
        options: [.caseInsensitive]
      )
    ),
    Pattern(
      category: "api_key",
      regex: try! NSRegularExpression(
        pattern: #"\b(?:sk[-_][A-Za-z0-9_-]{16,}|ghp_[A-Za-z0-9_]{20,}|AKIA[0-9A-Z]{16})\b"#,
        options: []
      )
    ),
    Pattern(
      category: "one_time_code",
      regex: try! NSRegularExpression(
        pattern: #"\b(?:one[- ]?time code|verification code|2fa code|mfa code|otp)\s+(?:is\s+)?['"]?([0-9]{4,8})\b"#,
        options: [.caseInsensitive]
      )
    ),
    Pattern(
      category: "token",
      regex: try! NSRegularExpression(
        pattern: #"\b(?:token|auth[_-]?token|access[_-]?token|bearer)\s*[:=]\s*['"]?([A-Za-z0-9._~+/=-]{16,})"#,
        options: [.caseInsensitive]
      )
    ),
    Pattern(
      category: "password",
      regex: try! NSRegularExpression(
        pattern: #"\b(?:password|passwd|pwd)\s*[:=]\s*['"]?([^'"\s]{8,})"#,
        options: [.caseInsensitive]
      )
    ),
    Pattern(
      category: "cookie",
      regex: try! NSRegularExpression(
        pattern: #"\b(?:session|cookie)\s*[:=]\s*['"]?([A-Za-z0-9._~+/=-]{16,})"#,
        options: [.caseInsensitive]
      )
    ),
  ]

  static func detections(in text: String) -> [HardSecretDetection] {
    let nsRange = NSRange(text.startIndex..<text.endIndex, in: text)
    var detections: [HardSecretDetection] = []
    for pattern in patterns {
      for match in pattern.regex.matches(in: text, options: [], range: nsRange) {
        guard let range = Range(match.range, in: text) else { continue }
        detections.append(HardSecretDetection(category: pattern.category, range: range))
      }
    }
    return detections
  }

  static func containsHardSecret(_ text: String) -> Bool {
    !detections(in: text).isEmpty
  }

  static func categories(in text: String) -> [String] {
    Array(Set(detections(in: text).map(\.category))).sorted()
  }
}
